### DISCUSSION TITLE
Webhooks?

### DISCUSSION THREAD

**Discussion by osseonews** [vendurehq/vendure#634] [Q&A]
Are there webhooks available in Vendure? For example, if there is a new order, can we have webhook hit an external endpoint to execute some code?

**Comment by michaelbromley** [comment]
Hi,
There are no webhooks by default, but you can implement them yourself in a plugin. The general idea would be to create a plugin which uses the [EventBus](https://www.vendure.io/docs/typescript-api/events/event-bus/) to listen for whichever [events](https://www.vendure.io/docs/typescript-api/events/event-types/) you are interested in and then make the http request to your desired endpoint. 

Rough sketch:

```TypeScript
import { OnVendureBootstrap, OrderStateTransitionEvent, PluginCommonModule, 
  VendurePlugin, EventBus } from '@vendure/core';

@VendurePlugin({
  imports: [PluginCommonModule],
})
export class WebhooksPlugin implements OnVendureBootstrap {

  constructor(
    private workerService: WorkerService,
    private eventBus: EventBus,
  ) {}
  
  /**
   * When the server bootstraps, set up a subscription for events 
   * published whenever an Order changes state. 
   */
  onVendureBootstrap() {
    this.eventBus.ofType(OrderStateTransitionEvent).subscribe(event => {
      if (event.toState === 'PaymentSettled') {
        // Here you make your webhook request.
        // You could even make the webhook urls configurable using
        // the `init` pattern demonstrated here: 
        // https://github.com/vendure-ecommerce/plugin-template/blob/db7a0aa415775a8352dcb3e7d373a56e2f6cb3d6/src/plugin.ts#L51-L58
      }
    });
  }

}
```

**Comment by martijnvdbrug** [comment]
I have written a plugin that triggers a channel aware webhook on configurable events for my project: [github source](https://github.com/martijnvdbrug/shops/tree/develop/vendure/src/webhook)
~~I have yet to publish it to NPM, but for now you can just copy the `vendure/src/webhook` directory and follow the readme.~~
Here is the NPM package: https://www.npmjs.com/package/vendure-plugin-webhook
